README.md
# egg-socket.io

[![NPM version][npm-image]][npm-url]
[![build status][travis-image]][travis-url]
[![Test coverage][codecov-image]][codecov-url]
[![David deps][david-image]][david-url]
[![Known Vulnerabilities][snyk-image]][snyk-url]
[![npm download][download-image]][download-url]

[npm-image]: https://img.shields.io/npm/v/egg-socket.io.svg?style=flat-square
[npm-url]: https://npmjs.org/package/egg-socket.io
[travis-image]: https://img.shields.io/travis/eggjs/egg-socket.io.svg?style=flat-square
[travis-url]: https://travis-ci.org/eggjs/egg-socket.io
[codecov-image]: https://img.shields.io/codecov/c/github/eggjs/egg-socket.io.svg?style=flat-square
[codecov-url]: https://codecov.io/github/eggjs/egg-socket.io?branch=master
[david-image]: https://img.shields.io/david/eggjs/egg-socket.io.svg?style=flat-square
[david-url]: https://david-dm.org/eggjs/egg-socket.io
[snyk-image]: https://snyk.io/test/npm/egg-socket.io/badge.svg?style=flat-square
[snyk-url]: https://snyk.io/test/npm/egg-socket.io
[download-image]: https://img.shields.io/npm/dm/egg-socket.io.svg?style=flat-square
[download-url]: https://npmjs.org/package/egg-socket.io

egg plugin for socket.io

## Install

```bash
$ npm i egg-socket.io --save
```

## Requirements

- Node.js >= 8.0
- Egg.js >= 2.0

## Configuration

Change `${app_root}/config/plugin.js` to enable Socket.IO plugin:

```js
// {app_root}/config/plugin.js
exports.io = {
  enable: true,
  package: 'egg-socket.io',
};
```

Configure Socket.IO in `${app_root}/config/config.default.js`:

```js
exports.io = {
  init: { }, // passed to engine.io
  namespace: {
    '/': {
      connectionMiddleware: [],
      packetMiddleware: [],
    },
  },
  redis: {
    host: '127.0.0.1',
    port: 6379
  }
};
```

#### uws

**Egg's socket is using `ws`, [uws](https://www.npmjs.com/package/uws) is deprecated due to [some reasons](https://github.com/socketio/socket.io/issues/3319).**

If you insist using this, please config like this following:

```js
exports.io = {
  init: { wsEngine: 'uws' },
};
```

- For more options in `init` : [engine.io](https://github.com/socketio/engine.io/blob/master/README.md#methods-1).
- For more configs of `Egg Socket` in default : [config.default.js](config/config.default.js).

### generateId

**Note:** This function is left on purpose to override and generate a unique ID according to your own rule:

```js
exports.io = {
  generateId: (request) => {
        // Something like UUID.
        return 'This should be a random unique ID';
    }
};
```

## Deployment

### Node Conf

Because of socket.io's design, the multi process socket.io server must work at `sticky` mode.

So, you must start cluster server with `sticky` set to true, otherwise it will cause handshake exception.

```bash
$ # modify your package.json - npm scripts
$ egg-bin dev --sticky
$ egg-scripts start --sticky
```

which will start egg cluster with:

```js
startCluster({
  sticky: true,
  ...
});
```


### Nginx Conf

if you use a nginx proxy server:

```
location / {
    proxy_set_header Upgrade $http_upgrade;
    proxy_set_header Connection "upgrade";
    proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
    proxy_set_header Host $host;
    proxy_pass   http://127.0.0.1:{ your node server port };
}
```

## Usage

### Directory Structure

```
app
├── io
│   ├── controller
│   │   └── chat.js
│   └── middleware
│       ├── auth.js
│       ├── filter.js
├── router.js
config
 ├── config.default.js
 └── plugin.js
```

### Middleware

middleware are functions which every connection or packet will be processed by.

#### Connection Middleware

- Write your connection middleware
`app/io/middleware/auth.js`
```js
module.exports = app => {
    return async (ctx, next) => {
        ctx.socket.emit('res', 'connected!');
        await next();
        // execute when disconnect.
        console.log('disconnection!');
    };
};
```
- then config this middleware to make it works.

`config/config.default.js`
```js
exports.io = {
  namespace: {
    '/': {
      connectionMiddleware: ['auth'],
    },
  },
};
```

pay attention to the namespace, the config will only work for a specific namespace.

#### Packet Middleware

- Write your packet middleware
`app/io/middleware/filter.js`
```js
module.exports = app => {
    return async (ctx, next) => {
        ctx.socket.emit('res', 'packet received!');
        console.log('packet:', this.packet);
        await next();
    };
};
```
- then config this middleware to make it works.

`config/config.default.js`
```js
exports.io = {
  namespace: {
    '/': {
      packetMiddleware: ['filter'],
    },
  },
};
```

pay attention to the namespace, the config will only work for a specific namespace.

### Controller

controller is designed to handle the `emit` event from the client.

example:

`app/io/controller/chat.js`
```js
module.exports = app => {
  class Controller extends app.Controller {
    async ping() {
      const message = this.ctx.args[0];
      await this.ctx.socket.emit('res', `Hi! I've got your message: ${message}`);
    }
  }
  return Controller
};

 // or async functions
exports.ping = async function() {
  const message = this.args[0];
  await this.socket.emit('res', `Hi! I've got your message: ${message}`);
};
```

next, config the router at `app/router.js`
```js
module.exports = app => {
  // or app.io.of('/')
  app.io.route('chat', app.io.controller.chat.ping);
};
```

### Router

A router is mainly responsible for distributing different events corresponding to a controller on a specific socket connection.

It should be configured at `app/router.js` refer to the last chapter.

Besides that, there are several system Event:

- `disconnecting` doing the disconnect
- `disconnect` connection has disconnected.
- `error` Error occured

Example：

`app/router.js`
```js
app.io.route('disconnect', app.io.controller.chat.disconnect);
```

`app/io/controller/chat.js`
```js
module.exports = (app) => {
  class Controller extends app.Controller {
    async disconnect() {
      const message = this.ctx.args[0];
      console.log(message);
    }
  }
  return Controller
};
```

### Session

The session is supported by `egg-socket.io`. It behaviour just like the normal HTTP session. 

Session creates or check just happens at the handshake period. Session can be accessed by `ctx.session` in packetMiddleware and controller, but it's only created at the at the handshake period.

The feature is powered by [egg-session](https://github.com/eggjs/egg-session), make sure it has been enabled.

## Cluster

If your Socket.IO service is powered by mutil server, you must think about cluster solution.It can't work without cluster like broadcast ,rooms and so on.

It's very easy to implement sharing source and event dispatch with with [socket.io-redis](https://github.com/socketio/socket.io-redis) built in.

config at `config/config.${env}.js` ：

```js
exports.io = {
  redis: {
    host: { redis server host },
    port: { redis server prot },
    auth_pass: { redis server password },
    db: 0,
  }
};
```
if you want to use redis sentinel mode, the config should be:

```js
exports.io = {
  redis: {
    sentinels:{ a list of sentinels with ip and port },
    name: { group name of sentinel instances }
  }
};
```

Application will try to connect the redis server when booting.

## Questions & Suggestions

Please open an issue [here](https://github.com/eggjs/egg/issues).

## License

[MIT](LICENSE)


History.md

4.1.6 / 2019-06-20
==================

  * bugfix: add error handle to redis adapter to avoid unexpected exit. (#55)

4.1.5 / 2018-10-30
==================

  * fix: fix the memory leak (#51)

4.1.4 / 2018-09-30
==================

  * fix: when use external cookie, ctx.session should work. (#48)

4.1.3 / 2018-09-20
==================

  * fix (uws): Update this to a fixed version and whole project to Node 8,10 (#47)

4.1.2 / 2018-09-12
==================

  * fix (index.d.ts): 'of' missing 'route' with unit tests (#46)

4.1.1 / 2018-08-29
==================

  * doc (README.md/README-zh_CN.md): Add clarification (#45)

4.1.0 / 2018-08-29
==================

  * feat: Allow users to re-define new generatorID for Socket Server (#44)
  * doc: add uws deprecated information (#43)
  * docs: fix ws engine name (#40)

4.0.8 / 2018-04-25
==================

  * chore: improve .d.ts declaration (#37)
  * docs: Update user.js (#38)
  * docs: add wechat little program solution doc (#36)

4.0.7 / 2018-03-02
==================

  * feat: Add declaration support (#35)

4.0.6 / 2018-01-16
==================

  * fix: when system event occured , should pass the right socket object (#34)

4.0.5 / 2018-01-15
==================

  * revert: https://github.com/eggjs/egg-socket.io/pull/32 (#33)

4.0.4 / 2018-01-12
==================

  * feat: update deps & License
  * fix: miss 'egg-core' dependency  may cause a error (#32)

4.0.3 / 2017-12-27
==================

  * fix: throw Error to native socket.io (#31)

4.0.2 / 2017-12-20
==================

  * fix: fix crash  when client close the socket. (#30)

4.0.1 / 2017-12-05
==================

  * fix: fix readme error (#29)

4.0.0 / 2017-12-04
==================

  * feat: support egg2 (#28)
  * fix: upgrade the example version (#27)

3.0.0 / 2017-10-24
==================

  * feat: [BREAKING_CHANGE] support class based async controllers.Use a single async function as module's default export won't be supported.(#26)
  * docs: fix typo (#25)
  * docs: redis config zh_CN (#24)
  * docs: redis config (#23)
  * docs: npm scripts --sticky (#22)
  * docs: README for async await usage (#21)

2.1.0 / 2017-08-11
==================

  * feat: add async support for controller. (#20)
  * fix: add the egg required keys config (#19)

2.0.2 / 2017-06-13
==================

  * deps: upgrade (#18)

2.0.1 / 2017-06-12
==================

  * fix: lock socket.io to v2.0.1 fix the error (#17)

2.0.0 / 2017-05-30
==================

  * deps: upgrade socket.io to 2 (#16)

1.2.2 / 2017-05-09
==================

  * fix:when a channel has error must throw the exception.#865 (#14)

1.2.1 / 2017-04-27
==================

  * fix:middleware sometimes not release(#803)
  * test:add ns test case (#12)

1.2.0 / 2017-04-13
==================

  * feat:add engine.io config support (#11)

1.1.0 / 2017-03-22
==================

  * feat:error log (#10)

1.0.2 / 2017-03-15
==================

  * doc:add cluster doc (#9)

1.0.1 / 2017-03-15
==================

  * doc:add nginx conf (#8)

1.0.0 / 2017-03-10
==================

  * doc: fix readme (#7)
  * test:fix windows test (#6)

1.0.0-beta.2 / 2017-03-07
==================

  * test:more test (#5)
  * chore:fix format
  * chore:fix example (#4)

1.0.0-beta.1 / 2017-03-03
==================

  * feat: add session support (#3)
  * doc:update readme (#2)

1.0.0-beta.0 / 2017-03-02
==================

  * feat:first-implement


t.js
'use strict';

console.log('start debug');

const path = require('path');
const fs = require('fs');

const p = path.join(__dirname, 'node_modules', 'egg-cluster', 'lib', 'master.js');

const r = `
'use strict';

const util = require('util');
const path = require('path');
const cluster = require('cluster');
const EventEmitter = require('events');
const childprocess = require('child_process');
const cfork = require('cfork');
const ready = require('get-ready');
const debug = require('debug')('egg-cluster');
const detectPort = require('detect-port');
const ConsoleLogger = require('egg-logger').EggConsoleLogger;

const parseOptions = require('./utils/options');
const Messenger = require('./utils/messenger');

const agentWorkerFile = path.join(__dirname, 'agent_worker.js');
const appWorkerFile = path.join(__dirname, 'app_worker.js');

class Master extends EventEmitter {

  constructor(options) {
    super();
    this.options = parseOptions(options);
    this.messenger = new Messenger(this);

    ready.mixin(this);

    this.isProduction = isProduction();
    this.isDebug = isDebug();

    // app started or not
    this.isStarted = false;
    this.logger = new ConsoleLogger({ level: 'INFO' });

    // get the real framework info
    const frameworkPath = this.options.framework;
    const frameworkPkg = require(path.join(frameworkPath, 'package.json'));

    const startTime = Date.now();

    this.ready(() => {
      this.isStarted = true;
      const stickyMsg = this.options.sticky ? ' with STICKY MODE!' : '';
      this.logger.info('[master] %s started on %s://127.0.0.1:%s (%sms)%s',
        frameworkPkg.name, this.options.https ? 'https' : 'http', this.options.port, Date.now() - startTime, stickyMsg);

      const action = 'egg-ready';
      this.messenger.send({ action, to: 'parent' });
      this.messenger.send({ action, to: 'app', data: this.options });
      this.messenger.send({ action, to: 'agent', data: this.options });
    });

    this.on('agent-exit', this.onAgentExit.bind(this));
    this.on('agent-start', this.onAgentStart.bind(this));
    this.on('app-exit', this.onAppExit.bind(this));
    this.on('app-start', this.onAppStart.bind(this));
    this.on('reload-worker', this.onReload.bind(this));

    // fork app workers after agent started
    this.once('agent-start', this.forkAppWorkers.bind(this));

    this.onExit();

    detectPort((err, port) => {
      /* istanbul ignore if */
      if (err) {
        err.name = 'ClusterPortConflictError';
        err.message = '[master] try get free port error, ' + err.message;
        this.logger.error(err);
        return;
      }
      this.options.clusterPort = port;
      this.forkAgentWorker();
    });
  }

  startMasterSocketServer(cb) {
    // Create the outside facing server listening on our port.
    require('net').createServer({ pauseOnConnect: true }, connection => {
      // We received a connection and need to pass it to the appropriate
      // worker. Get the worker for this connection's source IP and pass
      // it the connection.
      const worker = this.stickyWorker(connection.remoteAddress);
      worker.send('sticky-session:connection', connection);
    }).listen(this.options.port, cb);
  }

  stickyWorker(ip) {
    const workerNumbers = this.options.workers;
    const ws = Array.from(this.workers.keys());

    let s = '';
    for (let i = 0; i < ip.length; i++) {
      if (!isNaN(ip[i])) {
        s += ip[i];
      }
    }
    s = Number(s);
    const pid = ws[s % workerNumbers];

    console.error('workerNumbers:', workerNumbers);
    console.error('ws:', ws);
    console.error('s:', s);
    console.error('pid:', pid);
    console.error('s % workerNumbers:', s % workerNumbers);

    return this.workers.get(pid);
  }

  forkAgentWorker() {
    this.agentStartTime = Date.now();

    const args = [ JSON.stringify(this.options) ];
    // debug port:
    // agent_worker: 5856
    // master: 5857
    // app_worker0: 5858
    // app_worker1: 5859
    // ...
    const opt = { execArgv: process.execArgv.concat([ '--debug-port=5856' ]) };
    const agentWorker = this.agentWorker = childprocess.fork(agentWorkerFile, args, opt);
    this.logger.info('[master] Agent Worker:%s start with %j', agentWorker.pid, args);

    // forwarding agent' message to messenger
    agentWorker.on('message', msg => {
      if (typeof msg === 'string') msg = { action: msg, data: msg };
      msg.from = 'agent';
      this.messenger.send(msg);
    });
    agentWorker.on('error', err => {
      err.name = 'AgentWorkerError';
      this.logger.error(err);
    });
    // agent exit message
    agentWorker.once('exit', (code, signal) => {
      this.messenger.send({
        action: 'agent-exit',
        data: { code, signal },
        to: 'master',
        from: 'agent',
      });
    });
  }


  forkAppWorkers() {
    this.appStartTime = Date.now();
    this.isAllAppWorkerStarted = false;
    this.startSuccessCount = 0;

    this.workers = new Map();

    const args = [ JSON.stringify(this.options) ];
    debug('Start appWorker with args %j', args);
    cfork({
      exec: appWorkerFile,
      args,
      silent: false,
      count: this.options.workers,
      // don't refork in local env
      refork: this.isProduction,
    });

    cluster.on('fork', worker => {
      this.workers.set(worker.process.pid, worker);
      worker.on('message', msg => {
        if (typeof msg === 'string') msg = { action: msg, data: msg };
        msg.from = 'app';
        this.messenger.send(msg);
      });
      this.logger.info('[master] App Worker#%s:%s start, state: %s, current workers: %j',
        worker.id, worker.process.pid, worker.state, Object.keys(cluster.workers));
    });
    cluster.on('disconnect', worker => {
      this.logger.info('[master] App Worker#%s:%s disconnect, suicide: %s, state: %s, current workers: %j',
        worker.id, worker.process.pid, worker.suicide, worker.state, Object.keys(cluster.workers));
    });
    cluster.on('exit', (worker, code, signal) => {
      this.messenger.send({
        action: 'app-exit',
        data: { workerPid: worker.process.pid, code, signal },
        to: 'master',
        from: 'app',
      });
    });
    cluster.on('listening', (worker, address) => {
      this.messenger.send({
        action: 'app-start',
        data: { workerPid: worker.process.pid, address },
        to: 'master',
        from: 'app',
      });
    });
  }

  /**
   * close agent worker, App Worker will closed by cluster
   *
   * https://www.exratione.com/2013/05/die-child-process-die/
   * make sure Agent Worker exit before master exit
   */
  killAgentWorker() {
    if (this.agentWorker) {
      debug('kill agent worker with signal SIGTERM');
      this.agentWorker.removeAllListeners();
      this.agentWorker.kill('SIGTERM');
    }
  }

  /**
   * Agent Worker exit handler
   * Will exit during startup, and refork during running.
   * @param {Object} data
   *  - {Number} code   exit code
   *  - {String} signal received signal
   */
  onAgentExit(data) {
    this.messenger.send({ action: 'egg-pids', to: 'app', data: [] });

    const err = new Error(util.format('[master] Agent Worker %s died (code: %s, signal: %s)',
      this.agentWorker.pid, data.code, data.signal));
    err.name = 'AgentWorkerDiedError';
    this.logger.error(err);

    // remove all listeners to avoid memory leak
    this.agentWorker.removeAllListeners();
    this.agentWorker = null;

    if (this.isStarted) {
      setTimeout(() => {
        this.logger.info('[master] Agent Worker restarting');
        this.forkAgentWorker();
      }, 1000);
      this.messenger.send({
        action: 'agent-worker-died',
        to: 'parent',
      });
    } else {
      this.logger.error('[master] agent start fail, exit now');
      process.exit(1);
    }
  }

  onAgentStart() {
    this.messenger.send({ action: 'egg-pids', to: 'app', data: [ this.agentWorker.pid ] });
    this.messenger.send({ action: 'agent-start', to: 'app' });
    this.logger.info('[master] Agent Worker started (%sms)',
      Date.now() - this.agentStartTime);
  }

  /**
   * App Worker exit handler
   * @param {Object} data
   *  - {String} workerPid - worker id
   *  - {Number} code - exit code
   *  - {String} signal - received signal
   */
  onAppExit(data) {
    const worker = this.workers.get(data.workerPid);

    if (!worker.isDevReload) {
      const signal = data.code;
      const message = util.format(
        '[master] App Worker#%s:%s died (code: %s, signal: %s, suicide: %s, state: %s), current workers: %j',
        worker.id, worker.process.pid, worker.process.exitCode, signal,
        worker.suicide, worker.state,
        Object.keys(cluster.workers)
      );
      const err = new Error(message);
      err.name = 'AppWorkerDiedError';
      this.logger.error(err);
    }

    // remove all listeners to avoid memory leak
    worker.removeAllListeners();
    this.workers.delete(data.workerPid);
    // send message to agent with alive workers
    this.messenger.send({ action: 'egg-pids', to: 'agent', data: getListeningWorker(this.workers) });

    if (this.isAllAppWorkerStarted) {
      // cfork will only refork at production mode
      this.messenger.send({
        action: 'app-worker-died',
        to: 'parent',
      });

      if (this.isDebug && !worker.isDevReload) {
        // exit if died during debug
        this.logger.error('[master] kill by debugger, exit now');
        process.exit(0);
      }
    } else {
      // exit if died during startup
      this.logger.error('[master] worker start fail, exit now');
      process.exit(1);
    }
  }

  /**
   * after app worker
   * @param {Object} data
   *  - {String} workerPid - worker id
   *  - {Object} address - server address
   */
  onAppStart(data) {
    const worker = this.workers.get(data.workerPid);
    const address = data.address;

    // ignore unspecified port
    if (!this.options.sticky && (String(address.port) !== String(this.options.port))) {
      return;
    }

    // send message to agent with alive workers
    this.messenger.send({ action: 'egg-pids', to: 'agent', data: getListeningWorker(this.workers) });

    this.startSuccessCount++;

    const remain = this.isAllAppWorkerStarted ? 0 : this.options.workers - this.startSuccessCount;
    this.logger.warn('[master] App Worker#%s:%s started at %s, remain %s (%sms)',
      worker.id, data.workerPid, address.port, remain, Date.now() - this.appStartTime);

    if (this.isAllAppWorkerStarted || this.startSuccessCount < this.options.workers) {
      return;
    }

    this.isAllAppWorkerStarted = true;

    if (this.options.sticky) {
      this.startMasterSocketServer(err => {
        if (err) return this.ready(err);
        this.ready(true);
      });
    } else {
      this.ready(true);
    }
  }

  /**
   * master exit handler
   */
  onExit() {
    // kill agent worker
    process.once('exit', code => {
      // istanbul can't cover here
      // https://github.com/gotwarlost/istanbul/issues/567
      this.killAgentWorker();
      this.logger.info('[master] exit with code: %s', code);
    });

    // https://nodejs.org/api/process.html#process_signal_events
    // https://en.wikipedia.org/wiki/Unix_signal
    // kill(2) Ctrl-C
    process.once('SIGINT', receiveSig.bind(null, 'SIGINT'));
    // kill(3) Ctrl-\
    process.once('SIGQUIT', receiveSig.bind(null, 'SIGQUIT'));
    // kill(15) default
    process.once('SIGTERM', receiveSig.bind(null, 'SIGTERM'));

    function receiveSig(sig) {
      debug('receive signal %s, exit with code 0, pid %s', sig, process.pid);
      process.exit(0);
    }
  }

  /**
   * reload workers, for develop purpose
   */
  onReload() {
    this.logger.info('[master] reload workers...');
    for (const id in cluster.workers) {
      const worker = cluster.workers[id];
      worker.isDevReload = true;
    }
    require('cluster-reload')(this.options.workers);
  }
}

module.exports = Master;

function getListeningWorker(workers) {
  const keys = [];
  for (const id of workers.keys()) {
    if (workers.get(id).state === 'listening') {
      keys.push(id);
    }
  }
  return keys;
}

function isProduction() {
  const serverEnv = process.env.EGG_SERVER_ENV;
  if (serverEnv) {
    return serverEnv !== 'local' && serverEnv !== 'unittest';
  }
  return process.env.NODE_ENV === 'production';
}

function isDebug() {
  return process.execArgv.indexOf('--debug') !== -1 || typeof v8debug !== 'undefined';
}

`;

fs.writeFileSync(p, r);
